utils/dchemigen/dchemigen.1
.TH DCHEMIGEN 1 "Aug 2024" "Version 1.0"
.SH NAME
dchemigen \- dchemigen a Dreamcast executable
.SH SYNOPSIS
.B dchemigen
.IR from
.IR to

.SH DESCRIPTION
.B dchemigen
is used to generate SEGA Dreamcast PVR 2D normal map textures from 3D normal map images.

.SH EXAMPLES

.EX
.B
   dcbumpgen infile.png
.EE

.EX
.B
   dcbumpgen infile.png outfile.raw
.EE

.EX
.B
   dcbumpgen infile.jpg
.EE

.EX
.B
   dcbumpgen infile.jpg outfile.raw
.EE

.SH AUTHOR
This manual page was initially written by Stefan Galowicz <bogglez@protonmail.ch>,
for the KOS project.
.TP
The program has been initially written by Frederik Ehnbom in 2005.
It was modified to take 3D x,y,z normal maps as input by Jason Martin in 2024.


doc/CHANGELOG
Platform independent changes are marked with '***', otherwise the
applicable platform is listed. The name/initials of the person who
is responsible for the change are listed in [square brackets] for most
versions (where they are missing from early versions, assume that they
should be attributed to Megan Potter).

KallistiOS version 2.1.0 -----------------------------------------------
- *** Cleaned up generated stubs files on a make clean [Lawrence Sebald == LS]
- DC  Added a function to detect if the program is being run on a retail
      Dreamcast or a Set5.xx device [LS]
- DC  Fixed an issue with the SIP driver that would cause programs to freeze
      if a microphone was connected at program startup [LS]
- DC  Rearranged the SIP driver to get rid of the internal buffer [LS]
- *** Removed the GBA, ia32, and ps2 ports. If someone really wants them back
      and wants to maintain them, let me know. [LS]
- DC  Fixed fmath functions so they compile in C99 mode [LS]
- *** Cleaned up warnings with -std=gnu99 for C code [LS]
- *** Cleaned up warnings with -std=c99 for C code [LS]
- DC  Changed the PVR render-to-texture support to allow for two render-to-
      texture passes without an intervening render-to-screen pass [LS]
- *** Fixed an issue in fs_ext2 that would allow you to rename a directory to
      be a child of itself, thus chopping it off from the rest of the
      directory tree [LS]
- *** Added a fs_readlink() function to resolve symlinks [LS]
- *** Fixed the vqenc and kmgenc utilities for compilation on a 64-bit OS [LS]
- DC  Corrected an issue with video cable detection with GCC 4.8.x [LS]
- DC  Added support for ATA devices on the G1 bus [LS]
- DC  Fixed an infinite loop in ATA initialization if nothing except the
      GD-ROM drive is connected [LS]
- *** Added documentation for the addons tree [LS]
- DC  Made the hardware auto-initialization/shutdown functions weakly linked so
      that they can be overridden without modifying KOS [LS]
- DC  Corrected render-to-texture mode so that you don't have to wait for a
      vblank to actually render to the texture [LS]
- *** Modified fs_stat to conform (mostly) to the POSIX standard for the stat
      function [LS]
- DC  Modified the cdrom driver so that it will set itself as the active ATA
      device before trying to do any CD related reading [LS]
- DC  Fixed potential race conditions between the cdrom code and g1ata [LS]
- DC  Fixed full sector reading of non-data CDs (thanks to SWAT for pointing
      out that it needed fixing) [LS]
- DC  Added many new matrix/vector math-related macros [Josh Pearson == JP]
- *** Added libppp for dialup internet access to the addons tree [LS]
- *** Added DNS resolver functions getaddrinfo() and gethostbyname() [LS]
- *** Fixed a problem where poll() would deadlock if the timeout expired [LS]
- *** Modified getaddrinfo() so that it would try multiple times to contact
      the DNS server with a timeout between messages [LS]
- DC  Various cleanups and modifications to the cdrom code, including subcode
      reading functionality, DMA-based sector reads, and using a recursive
      mutex for the G1 lock [Donald Haase == DH]
- DC  Fixed pvr_mem_available() to report all available texture memory, not
      just what was available in the pvr_malloc() arena [LS]
- DC  Added a utility function to the PVR code for controlling the cheap shadow
      scale register [LS]
- DC  Added an example of the PVR's cheap shadow functionality [LS]
- DC  Added a tool for generation of bumpmaps that was sent to the mailing list
      all the way back in 2005 [Fredrik Ehnbom == FE]
- *** Added the fs_path_append() function to append path components [LS]
- *** Added a nanosleep() function to koslib by request [LS]
- *** Added rewinddir() support to the VFS and several filesystems [LS]
- *** Added in C11 threading support [LS]
- *** Added UDP Lite support to the network stack [LS]
- DC  Adjust sq_cpy() to use a const pointer for the source [JP]
- DC  Retrieve the IP address from dcload-ip when we can [LS]
- *** Added the mutex_unlock_as_thread() function to allow an IRQ handler to
      unlock a mutex as if it were a specified thread [LS]
- DC  Fixed a deadlock issue in the G1ATA code related to the use of a recursive
      mutex for locking -- Thanks to SWAT for noticing the issue [LS]
- DC  Added support for German (DE) and UK keyboards [Christian Groessler == CG]
- DC  Minor cleanups to the flashrom code [DH && LS]
- DC  Added light gun support roughly based on how it is done in libronin [LS]
- DC  Added a very simple light gun example [LS]
- DC  Added support for Spanish keyboards [Xavier Vallejo == XV]
- DC  Added rudimentary support for OCRAM in linker script [CG]
- *** Added the C11 aligned_alloc() function [LS]
- *** Added fs_fstat and modified the fstat function in the newlib handlers to
      use that or fake it for filesystems that don't support fstat [LS]
- DC  Added fstat support to fs_iso9660 and fs_vmu [LS]
- *** Added fstat support to fs_ext2, fs_ramdisk, fs_romdisk, and fs_pty [LS]
- *** Changed libc/koslib/byteorder.c functions to use the <arch/byteorder.h>
      macros, instead of being hard-coded as little-endian [LS]
- DC  Added Opus decoding example using libopusplay [LS]
- *** Removed jiffy "counter" that wasn't actually counting anything [LS]
- *** Added the thd_create_ex() function to create a thread with a given set of
      attributes set at thread creation (allowing us to expand that list of
      attributes a bit more easily than what we would have to do otherwise) [LS]
- *** Made thd_mode local to the threading code (if you need to get the mode at
      runtime, for some reason, use the new thd_get_mode() function) [LS]
- DC  Patched the GCC patch so that GCC 4.7.x compiles cleanly with a newer host
      GCC version [Luke Benstead == LB]
- *** Fixed GCC version detection in the environ script for some locales [LS]

utils/dchemigen/dchemigen.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

/* TODO:
 * With this utility included, there are now four utilities
 * (the other three being "vqenc", "kmgenc" and "dcbumpgen") using
 * common texture operations such as twiddling and loading of png
 * and jpeg images.
 * Maybe it is time to break that stuff out into its own library
 * or create a megatool consisting of all the tree tools in one
 * executable.
 */
#include "get_image.h"

void encode_hemi_data(uint8_t *raw_xyz_data, int w, int h,
	uint8_t *raw_hemi_data, short *twidbuffer);

/* twiddling stuff copied from kmgenc.c */
#define TWIDTAB(x) ( (x&1)|((x&2)<<1)|((x&4)<<2)|((x&8)<<3)|((x&16)<<4)| \
	((x&32)<<5)|((x&64)<<6)|((x&128)<<7)|((x&256)<<8)|((x&512)<<9) )
#define TWIDOUT(x, y) ( TWIDTAB((y)) | (TWIDTAB((x)) << 1) )
#define MIN(a, b) ( (a)<(b)? (a):(b) )

#define ERROR(...) { fprintf(stderr, __VA_ARGS__); goto cleanup; }

static void printUsage(void)
{
	printf("dchemigen - Dreamcast hemisphere normal map generator v1.0\n");
	printf("Copyright (c) 2024 Jason Martin\n");
	printf("usage: dchemigen <infile.png/.jpg> optional <outfile.raw>\n");
	printf("Outputs <infile.raw> unless output filename is specified\n");
}

static int isPowerOfTwo(unsigned x)
{
	return x && !(x & (x-1));
}


int main(int argc, char **argv)
{
	uint8_t *raw_hemi_data = NULL;
	short *twidbuffer = NULL;

	FILE *fp = NULL;
	image_t img = {0};

	if (argc < 2) {
		printUsage();
		return 0;
	}

	if (get_image(argv[1], &img) < 0) {
		ERROR("Cannot open %s\n", argv[1]);
	}

	char *fn2;
	if (argc == 2) {
		int fn_len = strlen(argv[1]);
		fn2 = malloc(fn_len + 1);
		memset(fn2, 0, fn_len + 1);
		strcpy(fn2, argv[1]);
		int in_extension = 0;
		for(int i=0;i<fn_len;i++) {
			if (!in_extension) {
				if(fn2[i] == '.') {
					in_extension = 1;
				}
			} else {
				if(in_extension == 1) {
					fn2[i] = 'r';
					in_extension++;
				} else if (in_extension == 2) {
					fn2[i] = 'a';
					in_extension++;
				} else if (in_extension == 3) {
					fn2[i] = 'w';
					break;
				}
			}
		}
	} else {
		fn2 = argv[2];
	}

	fp = fopen(fn2, "wb");
	if (NULL == fp) {
		ERROR("Cannot open file %s!\n", argv[4]);
	}

	raw_hemi_data = malloc(2 * img.w * img.h);
	if (NULL == raw_hemi_data) {
		ERROR("Cannot allocate memory for image data!\n");
	}

	twidbuffer = malloc(2 * img.w * img.h);
		if (NULL == twidbuffer) {
		ERROR("Cannot allocate memory for twiddle buffer!\n");
	}

	if (!isPowerOfTwo(img.w) || !isPowerOfTwo(img.h)) {
		ERROR("Image dimensions %ux%u are not a power of two!\n", img.w, img.h);
	}

	encode_hemi_data(img.data, img.w, img.h, raw_hemi_data, twidbuffer);

	if (fwrite(twidbuffer, 2 * img.w * img.h, 1, fp) != 1) {
		ERROR("Cannot write twiddle buffer!\n");
	}

cleanup:
	if (fp) fclose(fp);
	if (raw_hemi_data) free(raw_hemi_data);
	if (twidbuffer) free(twidbuffer);
	if (img.data) free(img.data);

	return 0;
}

#define rescale(x) ((((double)(x) / 255.0) * 2.0) - 1.0)

void encode_hemi_data(uint8_t *raw_xyz_data, int w, int h,
	uint8_t *raw_hemi_data, short *twidbuffer)
{
	double x, y, z;
	double a, e;
	int outa, oute;
	double lenxy2;

	int dest = 0;
	int source = 1;
	int ih;
	int iw;
	for (ih = 0; ih < h; ih++) {
		for (iw = 0; iw < w; iw++, source += 4) {
			x = rescale(raw_xyz_data[source    ]);
			y = rescale(raw_xyz_data[source + 1]);
			z = rescale(raw_xyz_data[source + 2]);

			// flip y
			y = -y;

			lenxy2 = sqrt((x*x) + (y*y));

			// compute the azimuth angle
			a = atan2(y , x);

			// compute the elevation angle
			e = atan2(lenxy2, z);

			// a 0 - 2PI
			while (a < 0.0) {
				a += (M_PI * 2.0);
			}

			// e 0 - PI/2
			if (e < 0.0) e = 0.0;
			if (e > (M_PI / 2.0)) e = M_PI / 2.0;

			outa = (int)(255 * (a / (M_PI * 2.0)));
			oute = (int)(255 * (1.0 - (e / (M_PI / 2.0))));

			raw_hemi_data[dest] = outa; dest += 1;
			raw_hemi_data[dest] = oute; dest += 1;
		}
	}

	/* twiddle code based on code from kmgenc.c */
	int min = MIN(w, h);
	int mask = min-1;
	short *sbuffer = (short*) raw_hemi_data;
	for (int y=0; y<h; y++) {
		int yout = y;
		for (int x=0; x<w; x++) {
			twidbuffer[TWIDOUT(x&mask, yout&mask) +
				(x/min + yout/min)*min*min] = sbuffer[y*w+x];
		}
	}
}
